File: src/Elevation.cs
// Docked Console, a terminal docked to the edge of the Windows desktop.
// Copyright (C) 2026 Locke Werks
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
using System.Diagnostics;
using System.Runtime.InteropServices;

namespace DockedConsole;

internal enum ElevationPolicy
{
    /// <summary>Elevate only when the terminal we would embed is going to be elevated.</summary>
    Auto,

    /// <summary>Never elevate. Report and stop if the terminal turns out to be elevated.</summary>
    Never,

    /// <summary>Always elevate at startup.</summary>
    Always,
}

/// <summary>
/// Integrity level matching.
///
/// Embedding works by reparenting the terminal's window into the dock, and UIPI forbids a
/// medium-integrity process from calling SetParent on a window owned by an elevated one.
/// That is not hypothetical here: `wt -w new` is serviced by whichever Windows Terminal
/// "monarch" process already exists, so if ANY elevated terminal is running, the window we
/// are handed belongs to it and an unelevated dock cannot touch it. SetParent fails, the
/// terminal is left floating, and the dock has nothing to show.
///
/// So the dock matches the terminal instead of fighting it: if the window is going to be
/// elevated, relaunch elevated first. On a machine with no elevated terminal, nothing
/// elevates and no prompt appears.
/// </summary>
internal static class Elevation
{
    private const string TerminalProcessName = "WindowsTerminal";

    /// <summary>Marks a relaunch so the new instance waits for the old one's mutex.</summary>
    public const string RelaunchFlag = "--elevated-relaunch";

    public static bool IsCurrentProcessElevated()
    {
        try
        {
            using var identity = System.Security.Principal.WindowsIdentity.GetCurrent();
            var principal = new System.Security.Principal.WindowsPrincipal(identity);
            return principal.IsInRole(System.Security.Principal.WindowsBuiltInRole.Administrator);
        }
        catch (Exception)
        {
            return false;
        }
    }

    /// <summary>Null when the process cannot be queried, which is not the same as false.</summary>
    public static bool? IsProcessElevated(int pid)
    {
        IntPtr process = Native.OpenProcess(Native.PROCESS_QUERY_LIMITED_INFORMATION, false, pid);
        if (process == IntPtr.Zero)
        {
            return null;
        }

        IntPtr token = IntPtr.Zero;
        IntPtr buffer = IntPtr.Zero;
        try
        {
            if (!Native.OpenProcessToken(process, Native.TOKEN_QUERY, out token))
            {
                return null;
            }

            buffer = Marshal.AllocHGlobal(sizeof(int));
            if (!Native.GetTokenInformation(token, Native.TokenElevation, buffer, sizeof(int), out _))
            {
                return null;
            }

            return Marshal.ReadInt32(buffer) != 0;
        }
        catch (Exception)
        {
            return null;
        }
        finally
        {
            if (buffer != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(buffer);
            }

            if (token != IntPtr.Zero)
            {
                Native.CloseHandle(token);
            }

            Native.CloseHandle(process);
        }
    }

    /// <summary>
    /// True when a Windows Terminal process is running elevated. Any such process can be
    /// the monarch that services our window request, which would hand us a window we
    /// cannot reparent unless we elevate too.
    /// </summary>
    public static bool AnyElevatedTerminalRunning()
    {
        try
        {
            foreach (var proc in Process.GetProcessesByName(TerminalProcessName))
            {
                using (proc)
                {
                    if (IsProcessElevated(proc.Id) == true)
                    {
                        return true;
                    }
                }
            }
        }
        catch (Exception ex)
        {
            Debug.WriteLine($"terminal elevation scan failed: {ex.Message}");
        }

        return false;
    }

    public static bool ShouldElevate(DockConfig cfg)
    {
        if (cfg.ParsedTerminal != TerminalKind.WindowsTerminal)
        {
            // conhost is launched by us and inherits our integrity level, so there is
            // never a mismatch to correct.
            return false;
        }

        if (IsCurrentProcessElevated())
        {
            return false;
        }

        return cfg.ParsedElevation switch
        {
            ElevationPolicy.Always => true,
            ElevationPolicy.Never => false,
            _ => AnyElevatedTerminalRunning(),
        };
    }

    /// <summary>
    /// True when the dock cannot work but the configuration forbids fixing it. Checked so
    /// the refusal happens before a terminal window is opened that could never be captured.
    /// </summary>
    public static bool BlockedByPolicy(DockConfig cfg) =>
        cfg.ParsedElevation == ElevationPolicy.Never
        && cfg.ParsedTerminal == TerminalKind.WindowsTerminal
        && !IsCurrentProcessElevated()
        && AnyElevatedTerminalRunning();

    /// <summary>
    /// Restarts this executable elevated, passing the original arguments through. Returns
    /// false if the user declined the prompt, which is a decision rather than an error.
    /// </summary>
    public static bool Relaunch(string[] args, out string message)
    {
        string? exe = Environment.ProcessPath;
        if (string.IsNullOrEmpty(exe))
        {
            message = "could not determine this executable's path";
            return false;
        }

        var startInfo = new ProcessStartInfo
        {
            FileName = exe,
            UseShellExecute = true,
            Verb = "runas",
        };

        foreach (var arg in args)
        {
            if (!string.Equals(arg, RelaunchFlag, StringComparison.OrdinalIgnoreCase))
            {
                startInfo.ArgumentList.Add(arg);
            }
        }

        startInfo.ArgumentList.Add(RelaunchFlag);

        try
        {
            Process.Start(startInfo);
            message = "relaunched elevated";
            return true;
        }
        catch (System.ComponentModel.Win32Exception ex) when (ex.NativeErrorCode == 1223)
        {
            message = "The elevation prompt was declined, so Docked Console did not start.";
            return false;
        }
        catch (Exception ex)
        {
            message = $"could not relaunch elevated: {ex.Message}";
            return false;
        }
    }
}

File: src/elevation.cpp
// Docked Console, a terminal docked to the edge of the Windows desktop.
// Copyright (C) 2026 Locke Werks
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
#include "elevation.h"

#include "config.h"
#include "win32.h"

#include <shellapi.h>
#include <tlhelp32.h>

namespace dock {
namespace {

bool TokenIsElevated(HANDLE process, bool& elevated)
{
    HANDLE raw_token = nullptr;
    if (!OpenProcessToken(process, TOKEN_QUERY, &raw_token)) {
        return false;
    }

    Handle token(raw_token);
    TOKEN_ELEVATION info{};
    DWORD returned = 0;
    if (!GetTokenInformation(token.Get(), TokenElevation, &info, sizeof(info),
                             &returned)) {
        return false;
    }

    elevated = info.TokenIsElevated != 0;
    return true;
}

/// Quotes one argument for a command line the shell will re-split. Only the
/// cases that can occur here are handled: a value containing spaces or quotes.
std::wstring QuoteArg(const std::wstring& arg)
{
    if (!arg.empty() && arg.find_first_of(L" \t\"") == std::wstring::npos) {
        return arg;
    }

    std::wstring out = L"\"";
    size_t backslashes = 0;
    for (const wchar_t c : arg) {
        if (c == L'\\') {
            ++backslashes;
            continue;
        }
        if (c == L'"') {
            // Backslashes before a quote must be doubled, then the quote escaped.
            out.append(backslashes * 2 + 1, L'\\');
            out.push_back(L'"');
        } else {
            out.append(backslashes, L'\\');
            out.push_back(c);
        }
        backslashes = 0;
    }
    out.append(backslashes * 2, L'\\');
    out.push_back(L'"');
    return out;
}

} // namespace

bool IsCurrentProcessElevated()
{
    bool elevated = false;
    return TokenIsElevated(GetCurrentProcess(), elevated) && elevated;
}

int IsProcessElevated(DWORD pid)
{
    Handle process(OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid));
    if (!process) {
        return -1;
    }

    bool elevated = false;
    if (!TokenIsElevated(process.Get(), elevated)) {
        return -1;
    }
    return elevated ? 1 : 0;
}

bool AnyElevatedTerminalRunning()
{
    Handle snapshot(CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0));
    if (!snapshot) {
        return false;
    }

    PROCESSENTRY32W entry{};
    entry.dwSize = sizeof(entry);
    if (!Process32FirstW(snapshot.Get(), &entry)) {
        return false;
    }

    do {
        if (EqualsNoCase(entry.szExeFile, L"WindowsTerminal.exe")
            && IsProcessElevated(entry.th32ProcessID) == 1) {
            return true;
        }
    } while (Process32NextW(snapshot.Get(), &entry));

    return false;
}

bool ShouldElevate(const Config& cfg)
{
    if (IsCurrentProcessElevated()) {
        return false;
    }

    switch (cfg.ParsedElevation()) {
    case ElevationPolicy::Always: return true;
    case ElevationPolicy::Never:  return false;
    case ElevationPolicy::Auto:   break;
    }
    return AnyElevatedTerminalRunning();
}

bool BlockedByPolicy(const Config& cfg)
{
    return cfg.ParsedElevation() == ElevationPolicy::Never
        && !IsCurrentProcessElevated()
        && AnyElevatedTerminalRunning();
}

bool Relaunch(const std::vector<std::wstring>& args, std::wstring& message)
{
    const std::wstring exe = ExePath();
    if (exe.empty()) {
        message = L"could not determine this executable's path";
        return false;
    }

    std::wstring parameters;
    for (const auto& arg : args) {
        if (EqualsNoCase(arg, kRelaunchFlag)) {
            continue;
        }
        if (!parameters.empty()) {
            parameters.push_back(L' ');
        }
        parameters += QuoteArg(arg);
    }
    if (!parameters.empty()) {
        parameters.push_back(L' ');
    }
    parameters += kRelaunchFlag;

    SHELLEXECUTEINFOW info{};
    info.cbSize = sizeof(info);
    // SEE_MASK_NOASYNC because this process exits immediately afterwards.
    // Without it the shell may still be servicing the request when we die, and
    // the elevated copy never starts.
    info.fMask = SEE_MASK_NOASYNC | SEE_MASK_FLAG_NO_UI;
    info.lpVerb = L"runas";
    info.lpFile = exe.c_str();
    info.lpParameters = parameters.c_str();
    info.nShow = SW_SHOWNORMAL;

    if (ShellExecuteExW(&info)) {
        message = L"relaunched elevated";
        return true;
    }

    if (GetLastError() == ERROR_CANCELLED) {
        message = L"The elevation prompt was declined, so Docked Console did not start.";
        return false;
    }

    message = L"could not relaunch elevated (error "
              + std::to_wstring(GetLastError()) + L")";
    return false;
}

} // namespace dock
